### DISCUSSION TITLE
Combined menu for undo, upload, and download (scan for quests)

### DISCUSSION THREAD

**Discussion by smichel17** [streetcomplete/StreetComplete#3330] [Ideas]
Since https://github.com/streetcomplete/StreetComplete/discussions/3251, I've been experimenting with manual upload, to save battery. This adds an upload button next to the main menu. At the same time, there is the undo button always floating in the bottom left.

When these buttons both had a single associated action, this made sense. However, now that the undo button opens up a menu anyway, it seems like a waste of space to have the extra upload button floating around, too. I presume it is not used that frequently (I seldom use it more than once or twice per session), so I think it would be fine to have in the menu.

While we're doing that, I think it would make sense to put "scan for quests here" in the same menu. And maybe also the "Add note" action, to make it more discoverable. So now this menu is basically for all the actions that interact with OSM data— upload, download, note, undo.

**Comment by mnalis** [comment]
Hmmmm, few things:
- moving UI elements around is **_generally a bad idea_**, unless there is a _very good_ reason for that (_and often even then_). People get used to where the things are, and are subsequently annoyed when then they have to re-train muscle memory. For example, I guess that more people switched to LibreOffice when Microsoft Office introduced then-new `ribbon` UI, then in any other time in history. It is not that new UI was bad (or at least, that was not the main reason :smile:), but because people were quite used to old UI, even if it required few more clicks than new one, that they found it less of a shock to switch to different program with UI similar to what they were used to, then to upgrade to new version of the same program, but with improved (ie. unfamiliar) UI.
- when one is not using manual upload, that icon does not exist currently, so it would have to be added always I guess (with different icon of course, or current users will never even try to click it if they think they know what is under that icon).
- undo issue 1: actually, I do use undo somewhat more extensively, especially when in area with spammy quests (like `house type`, `roof type`, etc). So what I do when I walk down such street is: when I am at the part of the street where I have some spare time, I map the all houses ahead as for example `house` - then I can concentrate on solving address numbers as I walk by. And then when I reach it and see that it is actually an garage and not single story house, I just undo it from the list, and mark it correctly. (I have to do it in that order, as I can't currently have `building type` be of lower priority that `address quest` [#2464](https://github.com/streetcomplete/StreetComplete/issues/2464))
- undo issue 2: I'd actually prefer if clicking current undo button it would automatically start multi-level undo, instead of having to jump through extra step. Now, if this was OsmAnd instead of StreetComplete, I'd be all too happy to suggest having the user choose which elements are shown on screen and where, so user could choose whether to display only multilevel undo icon, only singlelevel undo icon, or both. As it is not, I guess that is something I'd have fix on in my fork (by removing single-level undo option), when I find time. But moving it to other menu _at least would not add any clicks_ (but is still somewhat bad due to muscle-memory issue). I'd still prefer if the redundant functionality was just dropped and button was single-function again.
- having a button with many actions does make sense, though. Especially if it is going to be extended even more in the future. But can we not then get rid of even that button, and just extend standard hamburger button to take even bigger part of the screen? After all,  `scan for quests` and `team mode` are already there (and they seem to belong logically to that `multi-function` button).
- if we're going to either extend current hamburger button, or adding new multi-function button, I'd highly recommend adding `quick switch preset` item in there. That is the one I use often, and which is quite hard to get to (currently it is 9 steps to change it: `hamburger` -> `Settings` -> `Quest selection and order` -> `menu` -> `Manage presets` -> click new preset -> `back` -> `back` -> `back`). If it could become just `multi-function-icon` -> `switch preset` -> click new preset (which returned to map view by itself), it would be godsend.
- good idea about adding `Add note` there  to make it more discoverable. But hopefully people would realize they can still move it around after the note form pops up, or there might be new batch of complains about that.

**Reply by smichel17** [reply]
> moving UI elements around is **_generally a bad idea_**, unless there is a _very good_ reason for that (_and often even then_)

Agreed. I think freeing valuable screen real estate from being covered by a button (especially when a the quest form is open!) is a very good reason. I'm not sure if it is a good *enough* reason. I'm also not sure my proposed solution is the best one for the problem, which is (one reason) why I started a discussion instead of opening an issue.

> it would have to be added always I guess

I was thinking of making the new menu positioned in the bottom left. In other words, move the upload action into the undo menu and changing that button's icon. When auto-upload is enabled, the menu item would be hidden. I think this makes sense because the upload action should not be available when a quest form is open.

My biggest concern is that right now the upload button doubles as a big reminder, "You have manual upload enabled; don't forget to upload!" But I guess the little badge with the number of pending uploads could still be applied to the button in the new location.

> I do use undo somewhat more extensively

This sounds like a workaround for missing features— either [layers](https://github.com/streetcomplete/StreetComplete/issues/2461), a dedicated feature [for experienced mappers](https://github.com/streetcomplete/StreetComplete/discussions/3003) to show the address quest regardless of building type, or maybe just some way of [answering quests faster](https://github.com/streetcomplete/StreetComplete/issues/1735).

> But can we not then get rid of even that button, and just extend standard hamburger button to take even bigger part of the screen?

Although I've gotten used to it, I still don't like the hamburger menu opening a dialog; it's too large of a context switch. I can't find the place I mentioned it way back when the UI was updated to remove the top bar, but I'd prefer if the main menu would expand into a regular menu, like this:

[![fab expands to menu](https://raw.githubusercontent.com/gowong/material-sheet-fab/master/art/bgmonitor.gif)](https://github.com/gowong/material-sheet-fab)

But yes— while I wouldn't like it in the current form, we could definitely stick some or all of these actions into the main menu.

> `scan for quests` and `team mode` are already there (and they seem to belong logically to that `multi-function` button).

Hmm, I'm not sure about team mode. Upload and download both use bandwidth. Undo triggers a change to OSM data. They're actions with side-effects. Team mode, along with changing the current preset, feels more like a setting.

At least, that's how I was thinking about it. Maybe that's the wrong line to draw, and really it should just be "stuff that I'll want to use frequently" vs everything else.

Agreed that presets should probably be more prominent, although honestly I haven't used them much yet.

I don't recall seeing many bug reports about not being able to reposition the note when the button for adding a note was in the bottom left corner, and opened a similar interface to what we have today. So I'm not too worried about that.

**Comment by smichel17** [comment]
Note: I had this idea a few weeks ago, but the thing that reminded me to post it was reading https://github.com/streetcomplete/StreetComplete/pull/1478#issuecomment-929121853. It get me thinking, "What if we somehow combined the star counter with the main menu?" so that the intuitive thing is actually where the profile is.

**Comment by smichel17** [comment]
> I'd actually prefer if clicking current undo button it would automatically start multi-level undo

@mnalis Hmm, I think I agree. To undo the **most recent** answer with the current setup, you must:

1. Tap the undo button
2. Select "Undo last edit"
3. Select OK

If the history view were automatically opened, it would only add one step:

1. Tap the undo button
2. Select the most recent quest in the left sidebar or the map
3. Tap the quest icon with the :leftwards_arrow_with_hook: undo icon over it, in the left sidebar
4. Select OK

However, step 2 could be skipped by pre-selecting the most recent quest (maybe only if it was answered in the past 15 minutes).

If this is as easy to implement as I guess it will be, I think I'll try that and send a PR if I like it enough.

**Comment by smichel17** [comment]
https://github.com/streetcomplete/StreetComplete/pull/3349#issuecomment-943401116 is giving me ideas that this might still be a reasonable thing to do. There's already a "synced changes" element, so thematically it would make sense to have the button for syncing the rest of the changes nearby. …but, no concrete suggestions for the moment; one thing at a time.
